subiquity/common/tests/test_serialization.py

```diff
@@ -207,6 +219,61 @@ def test_serialize_union(self):
         }
         self.assertSerialization(typing.Union[Data, Container], data, expected)
 
+    def test_serialization_union_scalars(self):
+        self.assertSerialization(typing.Union[int, str], 30, 30)
+        self.assertSerialization(typing.Union[int, str], "hello", "hello")
+
+        self.assertSerialization(typing.Union[int, float], 30.0, 30.0)
+        self.assertSerialization(typing.Union[float, int], 30.0, 30.0)
+        self.assertSerialization(typing.Union[int, float], 30, 30)
+        self.assertSerialization(typing.Union[float, int], 30, 30)
+
+        self.assertSerialization(typing.Union[bool, str], True, True)
+        self.assertSerialization(typing.Union[bool, str], False, False)
+        self.assertSerialization(typing.Union[bool, str], "hello", "hello")
+
+        self.assertSerialization(typing.Union[int, float, str], 10, 10)
+        self.assertSerialization(typing.Union[int, float, str], "string", "string")
+        self.assertSerialization(typing.Union[int, float, str], 5.0, 5.0)
+
+        self.assertSerialization(typing.Union[int, float, str, None], 10, 10)
+        self.assertSerialization(typing.Union[int, float, str, None], None, None)
+
+        # This is what Optional[int] does.
+        self.assertSerialization(typing.Union[int, type(None)], 10, 10)
+        self.assertSerialization(typing.Union[int, type(None)], None, None)
+
+    def test_serialization_union_scalars__errors(self):
+        self.assertSerializationErrors(typing.Union[int, str], 5.0)
+        self.assertSerializationErrors(typing.Union[int, float], "hello")
+        self.assertSerializationErrors(typing.Union[int, float], "5.0")
+        self.assertSerializationErrors(typing.Union[str, int], True)
+
+    def test_serialization_union_none_attrs(self):
+        @attr.s(auto_attribs=True)
+        class A:
+            x: int
+
+        @attr.s(auto_attribs=True)
+        class B:
+            y: int
+
+        self.assertSerialization(typing.Union[type(None), A, B], None, None)
+        self.assertSerialization(
+            typing.Union[type(None), A, B], A(10), {"$type": "A", "x": 10}
+        )
+        self.assertSerialization(
+            typing.Union[type(None), A, B], B(10), {"$type": "B", "y": 10}
+        )
+
+        self.assertSerialization(typing.Union[type(None), A, B], None, None)
```

Comment:
This working is surprising to me. I need to revisit how this pans out. Does this work with non-attrs types?

Reply:
Oh nevermind, I see it now. Sorry!